### PR TITLE
feat(codecv1): support zstd compression

### PR DESCRIPTION
**Pull request by colinlyguo**
### Purpose or design rationale of this PR

Use https://github.com/DataDog/zstd as a base repo, and:

1. add `CompressScrollBatchBytes` [in this commit](https://github.com/colinlyguo/zstd/commit/d3f94bd13c765cbdd0e770ac827379eb3c2678e1).
2. to align with rust side, upgrade c zstd version to `v1.5.6` [in this commit](https://github.com/colinlyguo/zstd/commit/fcb551e6e516f5670ceb5a597ff1c5d06b6397b9), reusing scripts in: https://github.com/DataDog/zstd/pull/125.

- [ ] remove https://github.com/colinlyguo/zstd dependency, init and instead use a repo like https://github.com/scroll-tech/gozstd (changed `zstd` to `gozstd`, because scroll-tech repo already has a repo named https://github.com/scroll-tech/zstd).
- [ ] update rollup-relayer based on this PR.

### PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [x] feat: A new feature

### Deployment tag versioning

Has `tag` in `common/version.go` been updated or have you added `bump-version` label to this PR?

- [x] Yes

### Breaking change label

Does this PR have the `breaking-change` label?

- [x] No, this PR is not a breaking change

### CODE REVIEW

**Comment by colinlyguo** [comment]
Pretty weird that CI failed but is not related to this PR:
```
level=error msg="Running error: 1 error occurred:\n\t* can't run linter goanalysis_metalinter: buildir: failed to load package tracing: could not load export data: no export data for \"github.com/docker/compose/v2/internal/tracing\"\n\n"
lint.go:64: Error: Could not Lint error: exit status 3, cmd: /home/runner/work/scroll/scroll/common/build/bin/golangci-lint run --config ../build/.golangci.yml
exit status 1
make: *** [Makefile:7: lint] Error 1
```
looks familiar, will investigate this later.

File: common/types/encoding/codecv1/codecv1.go
Comment (noel2004) [inline common/types/encoding/codecv1/codecv1.go:13]: Maybe we should move this repo to scroll later?

**Review by noel2004 — APPROVED** [review]

**Comment by 0xmountaintop** [comment]
we should review https://github.com/colinlyguo/zstd/commit/d3f94bd13c765cbdd0e770ac827379eb3c2678e1 and https://github.com/colinlyguo/zstd/commit/fcb551e6e516f5670ceb5a597ff1c5d06b6397b9 first

**Comment by colinlyguo** [comment]
> we should review [colinlyguo/zstd@d3f94bd](https://github.com/colinlyguo/zstd/commit/d3f94bd13c765cbdd0e770ac827379eb3c2678e1) and [colinlyguo/zstd@fcb551e](https://github.com/colinlyguo/zstd/commit/fcb551e6e516f5670ceb5a597ff1c5d06b6397b9) first

we can initialize a new repo under scroll-tech (forked from https://github.com/DataDog/zstd), and I can move these commits to it, for review.

**Comment by colinlyguo** [comment]
discussed with @georgehao, we can package a zstd rust `.so` lib with functions like `compress_scroll_batch_bytes`, better supporting future zstd upgrades.

File: common/types/encoding/codecv1/codecv1.go
Comment (Thegaram) [inline common/types/encoding/codecv1/codecv1.go:208]: Just to make sure I understand correctly: when `withCompression = true`, the only things that change are: `blob`, `BlobVersionedHash`, right? So `z`, `DataHash` etc. won't change.

File: common/types/encoding/codecv1/codecv1.go
Comment (Thegaram) [inline common/types/encoding/codecv1/codecv1.go:13]: Do you plan to change this before merging this PR?

**Review by Thegaram — APPROVED** [review]

File: common/types/encoding/codecv1/codecv1.go
Comment (colinlyguo) [reply common/types/encoding/codecv1/codecv1.go:13]: Yes.

File: common/types/encoding/codecv1/codecv1.go
Comment (colinlyguo) [reply common/types/encoding/codecv1/codecv1.go:208]: Yes.

Comment (colinlyguo) [reply common/types/encoding/codecv1/codecv1.go:208]: `z` would be changed because `z` is influenced by `blobVersionedHash`.

**Comment by colinlyguo** [comment]
migrated to this repo: https://github.com/scroll-tech/da-codec
related PRs:
1. original codec (running in `mainnet` and `sepolia`): https://github.com/scroll-tech/da-codec/pull/1
2. added part: https://github.com/scroll-tech/da-codec/pull/2

**Comment by colinlyguo** [comment]
migrated to this repo: https://github.com/scroll-tech/da-codec
related PRs:
1. original codec (running in `mainnet` and `sepolia`): https://github.com/scroll-tech/da-codec/pull/1
2. added part: https://github.com/scroll-tech/da-codec/pull/2

**Pull request closed without merging**
